README
NAME
    Test::Number::Delta - Compare the difference between numbers against a
    given tolerance

SYNOPSIS
      # Import test functions
      use Test::Number::Delta;
      
  # Equality test with default tolerance
      delta_ok( 1e-5, 2e-5, 'values within 1e-6');
      
  # Inequality test with default tolerance
      delta_not_ok( 1e-5, 2e-5, 'values not within 1e-6');
      
  # Provide specific tolerance
      delta_within( 1e-3, 2e-3, 1e-4, 'values within 1e-4');         
      delta_not_within( 1e-3, 2e-3, 1e-4, 'values not within 1e-4');
      
  # Compare arrays or matrices
      @a = ( 3.14, 1.41 );
      @b = ( 3.15, 1.41 );
      delta_ok( \@a, \@b, 'compare @a and @b' );

      # Set a different default tolerance 
      use Test::Number::Delta within => 1e-5;
      delta_ok( 1.1e-5, 2e-5, 'values within 1e-5'); # ok
      
  # Set a relative tolerance
      use Test::Number::Delta relative => 1e-3;
      delta_ok( 1.01, 1.0099, 'values within 1.01e-3');

DESCRIPTION
    At some point or another, most programmers find they need to compare
    floating-point numbers for equality. The typical idiom is to test if the
    absolute value of the difference of the numbers is within a desired
    tolerance, usually called epsilon. This module provides such a function
    for use with Test::Harness. Usage is similar to other test functions
    described in Test::More. Semantically, the "delta_within" function
    replaces this kind of construct:

     ok ( abs($p - $q) < $epsilon, '$p is equal to $q' ) or
         diag "$p is not equal to $q to within $epsilon";

    While there's nothing wrong with that construct, it's painful to type it
    repeatedly in a test script. This module does the same thing with a
    single function call. The "delta_ok" function is similar, but either
    uses a global default value for epsilon or else calculates a 'relative'
    epsilon on the fly so that epsilon is scaled automatically to the size
    of the arguments to "delta_ok". Both functions are exported
    automatically.

    Because checking floating-point equality is not always reliable, it is
    not possible to check the 'equal to' boundary of 'less than or equal to
    epsilon'. Therefore, Test::Number::Delta only compares if the absolute
    value of the difference is less than epsilon (for equality tests) or
    greater than epsilon (for inequality tests).

USAGE
  use Test::Number::Delta;
    With no arguments, epsilon defaults to 1e-6. (An arbitrary choice on the
    author's part.)

  use Test::Number::Delta within => 1e-9;
    To specify a different default value for epsilon, provide a "within"
    parameter when importing the module. The value must be non-zero.

  use Test::Number::Delta relative => 1e-3;
    As an alternative to using a fixed value for epsilon, provide a
    "relative" parameter when importing the module. This signals that
    "delta_ok" should test equality with an epsilon that is scaled to the
    size of the arguments. Epsilon is calculated as the relative value times
    the absolute value of the argument with the greatest magnitude.
    Mathematically, for arguments 'x' and 'y':

     epsilon = relative * max( abs(x), abs(y) )

    For example, a relative value of "0.01" would mean that the arguments
    are equal if they differ by less than 1% of the larger of the two
    values. A relative value of 1e-6 means that the arguments must differ by
    less than 1 millionth of the larger value. The relative value must be
    non-zero.

  Combining with a test plan
     use Test::Number::Delta 'no_plan';
     
 # or
     
 use Test::Number::Delta within => 1e-9, tests => 1;

    If a test plan has not already been specified, the optional parameter
    for Test::Number::Delta may be followed with a test plan (see Test::More
    for details). If a parameter for Test::Number::Delta is given, it must
    come first.

FUNCTIONS
  delta_within
     delta_within(  $p,  $q, $epsilon, '$p and $q are equal within $epsilon' );
     delta_within( \@p, \@q, $epsilon, '@p and @q are equal within $epsilon' );

    This function tests for equality within a given value of epsilon. The
    test is true if the absolute value of the difference between $p and $q
    is less than epsilon. If the test is true, it prints an "OK" statement
    for use in testing. If the test is not true, this function prints a
    failure report and diagnostic. Epsilon must be non-zero.

    The values to compare may be scalars or references to arrays. If the
    values are references to arrays, the comparison is done pairwise for
    each index value of the array. The pairwise comparison is recursive, so
    matrices may be compared as well.

    For example, this code sample compares two matrices:

        my @a = (   [ 3.14, 6.28 ],
                    [ 1.41, 2.84 ]   );

        my @b = (   [ 3.14, 6.28 ],
                    [ 1.42, 2.84 ]   );

        delta_within( \@a, \@b, 1e-6, 'compare @a and @b' );

    The sample prints the following:

        not ok 1 - compare @a and @b
        # At [1][0]: 1.4100000 and 1.4200000 are not equal to within 0.000001

  delta_ok
     delta_ok(  $p,  $q, '$p and $q are close enough to equal' );
     delta_ok( \@p, \@q, '@p and @q are close enough to equal' );

    This function tests for equality within a default epsilon value. See
    "USAGE" for details on changing the default. Otherwise, this function
    works the same as "delta_within".

  delta_not_within
     delta_not_within(  $p,  $q, '$p and $q are different' );
     delta_not_within( \@p, \@q, $epsilon, '@p and @q are different' );

    This test compares inequality in excess of a given value of epsilon. The
    test is true if the absolute value of the difference between $p and $q
    is greater than epsilon. For array or matrix comparisons, the test is
    true if *any* pair of values differs by more than epsilon. Otherwise,
    this function works the same as "delta_within".

  delta_not_ok
     delta_not_ok(  $p,  $q, '$p and $q are different' );
     delta_not_ok( \@p, \@q, '@p and @q are different' );

    This function tests for inequality in excess of a default epsilon value.
    See "USAGE" for details on changing the default. Otherwise, this
    function works the same as "delta_not_within".

SEE ALSO
    Test::More, Test::Harness, Test::Builder

BUGS
    Please report any bugs or feature using the CPAN Request Tracker. Bugs
    can be submitted by email to "bug-Test-Number-Delta@rt.cpan.org" or
    through the web interface at
    <http://rt.cpan.org/Dist/Display.html?Queue=Test-Number-Delta>

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

AUTHOR
    David A Golden (DAGOLDEN)

    dagolden@cpan.org

    <http://dagolden.com/>

COPYRIGHT
    Copyright (c) 2005, 2006 by David A. Golden

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    The full text of the license can be found in the LICENSE file included
    with this module.

DISCLAIMER OF WARRANTY
    BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
    FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
    OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
    PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
    EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
    WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
    ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
    YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
    NECESSARY SERVICING, REPAIR, OR CORRECTION.

    IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
    WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
    REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE LIABLE
    TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL, OR
    CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE THE
    SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
    RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
    FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
    SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH
    DAMAGES.



Changes
Revision history for Perl module Test::Number::Delta

{{$NEXT}}

1.06      2014-11-25 07:09:08-05:00 America/New_York

    [Fixed]

    - Make 0 and 0 compare equal with a relative epsilon

    [Documented]

    - Added SEE ALSO with Number::Tolerant and Test::Deep::NumberTolerant

1.05      2014-11-24 11:08:11-05:00 America/New_York

    [Fixed]

    - relative comparison of array reference elements now works as intended

    - diagnostics of delta_not_ok and delta_not_within clarified under
      relative comparison

1.04      2013-11-20 18:43:32 America/New_York

    - Modernized distribution metadata and licensing

    - Distribution now managed with Dist::Zilla

1.03 Sun Sep  3 03:21:59 EDT 2006
    - fixed logic errors for delta_not_*
    - changed epsilon restriction to just non-zero; negative epsilons are
      the same as positive
    
1.02 Tue Aug 29 05:37:53 EDT 2006
    - added error checking to ensure that values of epsilon are always positive
    - added delta_not_within() and delta_not_ok() for inequality checking

1.01 Thu Jul 27 11:34:29 EDT 2006
    - changed comparison to less than epsilon, as checking "equal to"
      epsilon is, of course, not robust for floating point numbers
      
1.00 Tue Jul 25 18:19:51 EDT 2006
    - bumped version to 1.00 to indicate "release" status
    - added support for comparing arrays or matrices
    - bumped Test::Builder dependency to match my installed version

0.18 Sun Feb  5 06:48:15 EST 2006
    - removed Test::Exception as dependency
    - restored pod/pod-coverage tests as optional (skipped by default)
    - edited Pod for bugs/license tests
    - removed 5.006 syntax (warnings!) from test files

0.17 Mon Aug  1 10:55:13 EDT 2005
    - build_requires moved to requires to workaround CPANPLUS bug

0.16 Thu Jun  9 15:55:03 EDT 2005
    - removed pod/podcoverage testing to minimize dependencies

0.15 Tue Jun  7 17:46:36 EDT 2005
    - updated Build.PL for dependencies and traditional makefiles

0.14 Tue Mar  8 22:43:31 EST 2005
    - added pod and pod coverage testing

0.13 Wed Feb  2 22:06:56 EST 2005
    - added ability to pass through a test plan
    - added a parameter to specify that the default epsilon for delta_ok
      should be relative to the size of the arguments
    - updated pod in several areas for clarity and consistency

0.12 Mon Jan 24 23:45:21 EST 2005
    - pod fix
    - Build.PL cleanup

0.11 Tue Jan  4 00:01:20 EST 2005
    - renamed to Test::Number::Delta
    - renamed cmp_float() to delta_ok()
    - added delta_within()
    - added extra decimal place to numbers under comparison when printing
      diagnostics

0.10 Sun Dec 26 14:18:36 EST 2004
    - initial release as Test::Float

0.01 Sun Dec 26 13:12:10 2004
	- original version; created by ExtUtils::ModuleMaker::TT


lib/Test/Number/Delta.pm
use strict;
use warnings;

package Test::Number::Delta;
# ABSTRACT: Compare the difference between numbers against a given tolerance

our $VERSION = '1.07';

use vars qw (@EXPORT @ISA);

# Required modules
use Carp;
use Test::Builder;
use Exporter;

@ISA    = qw( Exporter );
@EXPORT = qw( delta_not_ok delta_ok delta_within delta_not_within );

=head1 SYNOPSIS

  # Import test functions
  use Test::Number::Delta;

  # Equality test with default tolerance
  delta_ok( 1e-5, 2e-5, 'values within 1e-6');

  # Inequality test with default tolerance
  delta_not_ok( 1e-5, 2e-5, 'values not within 1e-6');

  # Provide specific tolerance
  delta_within( 1e-3, 2e-3, 1e-4, 'values within 1e-4');
  delta_not_within( 1e-3, 2e-3, 1e-4, 'values not within 1e-4');

  # Compare arrays or matrices
  @a = ( 3.14, 1.41 );
  @b = ( 3.15, 1.41 );
  delta_ok( \@a, \@b, 'compare @a and @b' );

  # Set a different default tolerance
  use Test::Number::Delta within => 1e-5;
  delta_ok( 1.1e-5, 2e-5, 'values within 1e-5'); # ok

  # Set a relative tolerance
  use Test::Number::Delta relative => 1e-3;
  delta_ok( 1.01, 1.0099, 'values within 1.01e-3');


=head1 DESCRIPTION

At some point or another, most programmers find they need to compare
floating-point numbers for equality.  The typical idiom is to test
if the absolute value of the difference of the numbers is within a desired
tolerance, usually called epsilon.  This module provides such a function for use
with L<Test::More>.  Usage is similar to other test functions described in
L<Test::More>.  Semantically, the C<delta_within> function replaces this kind
of construct:

 ok ( abs($p - $q) < $epsilon, '$p is equal to $q' ) or
     diag "$p is not equal to $q to within $epsilon";

While there's nothing wrong with that construct, it's painful to type it
repeatedly in a test script.  This module does the same thing with a single
function call.  The C<delta_ok> function is similar, but either uses a global
default value for epsilon or else calculates a 'relative' epsilon on
the fly so that epsilon is scaled automatically to the size of the arguments to
C<delta_ok>.  Both functions are exported automatically.

Because checking floating-point equality is not always reliable, it is not
possible to check the 'equal to' boundary of 'less than or equal to
epsilon'.  Therefore, Test::Number::Delta only compares if the absolute value
of the difference is B<less than> epsilon (for equality tests) or
B<greater than> epsilon (for inequality tests).

=head1 USAGE

=head2 use Test::Number::Delta;

With no arguments, epsilon defaults to 1e-6. (An arbitrary choice on the
author's part.)

=head2 use Test::Number::Delta within => 1e-9;

To specify a different default value for epsilon, provide a C<within> parameter
when importing the module.  The value must be non-zero.

=head2 use Test::Number::Delta relative => 1e-3;

As an alternative to using a fixed value for epsilon, provide a C<relative>
parameter when importing the module.  This signals that C<delta_ok> should
test equality with an epsilon that is scaled to the size of the arguments.
Epsilon is calculated as the relative value times the absolute value
of the argument with the greatest magnitude.  Mathematically, for arguments
'x' and 'y':

 epsilon = relative * max( abs(x), abs(y) )

For example, a relative value of "0.01" would mean that the arguments are equal
if they differ by less than 1% of the larger of the two values.  A relative
value of 1e-6 means that the arguments must differ by less than 1 millionth
of the larger value.  The relative value must be non-zero.

=head2 Combining with a test plan

 use Test::Number::Delta 'no_plan';

 # or

 use Test::Number::Delta within => 1e-9, tests => 1;

If a test plan has not already been specified, the optional
parameter for Test::Number::Delta may be followed with a test plan (see
L<Test::More> for details).  If a parameter for Test::Number::Delta is
given, it must come first.

=cut

my $Test     = Test::Builder->new;
my $Epsilon  = 1e-6;
my $Relative = undef;

sub import {
    my $self  = shift;
    my $pack  = caller;
    my $found = grep /within|relative/, @_;
    croak "Can't specify more than one of 'within' or 'relative'"
      if $found > 1;
    if ($found) {
        my ( $param, $value ) = splice @_, 0, 2;
        croak "'$param' parameter must be non-zero"
          if $value == 0;
        if ( $param eq 'within' ) {
            $Epsilon = abs($value);
        }
        elsif ( $param eq 'relative' ) {
            $Relative = abs($value);
        }
        else {
            croak "Test::Number::Delta parameters must come first";
        }
    }
    $Test->exported_to($pack);
    $Test->plan(@_);
    $self->export_to_level( 1, $self, $_ ) for @EXPORT;
}

#--------------------------------------------------------------------------#
# _check -- recursive function to perform comparison
#--------------------------------------------------------------------------#

sub _check {
    my ( $p, $q, $e, $name, @indices ) = @_;
    my $epsilon;

    if ( !defined $e ) {
        $epsilon =
            $Relative
          ? $Relative * ( abs($p) > abs($q) ? abs($p) : abs($q) )
          : $Epsilon;
    }
    else {
        $epsilon = abs($e);
    }

    my ( $ok, $diag ) = ( 1, q{} ); # assume true
    if ( ref $p eq 'ARRAY' || ref $q eq 'ARRAY' ) {
        if ( @$p == @$q ) {
            for my $i ( 0 .. $#{$p} ) {
                my @new_indices;
                ( $ok, $diag, @new_indices ) =
                  _check( $p->[$i], $q->[$i], $e, $name, scalar @indices ? @indices : (), $i, );
                if ( not $ok ) {
                    @indices = @new_indices;
                    last;
                }
            }
        }
        else {
            $ok = 0;
            $diag =
                "Got an array of length "
              . scalar(@$p)
              . ", but expected an array of length "
              . scalar(@$q);
        }
    }
    else {
        $ok = $p == $q || abs( $p - $q ) < $epsilon;
        if ( !$ok ) {
            my ( $ep, $dp ) = _ep_dp($epsilon);
            $diag = sprintf( "%.${dp}f and %.${dp}f are not equal" . " to within %.${ep}f",
                $p, $q, $epsilon );
        }
    }
    return ( $ok, $diag, scalar(@indices) ? @indices : () );
}

sub _ep_dp {
    my $epsilon = shift;
    return ( 0, 0 ) unless $epsilon;
    $epsilon = abs($epsilon);
    my ($exp) = sprintf( "%e", $epsilon ) =~ m/e(.+)/;
    my $ep = $exp < 0 ? -$exp : 1;
    my $dp = $ep + 1;
    return ( $ep, $dp );
}

sub _diag_default {
    my ($ep) = _ep_dp( abs( $Relative || $Epsilon ) );
    my $diag = "Arguments are equal to within ";
    $diag .=
      $Relative
      ? sprintf( "relative tolerance %.${ep}f", abs($Relative) )
      : sprintf( "%.${ep}f",                    abs($Epsilon) );
    return $diag;
}

=head1 FUNCTIONS

=cut

#--------------------------------------------------------------------------#
# delta_within()
#--------------------------------------------------------------------------#

=head2 delta_within

 delta_within(  $p,  $q, $epsilon, '$p and $q are equal within $epsilon' );
 delta_within( \@p, \@q, $epsilon, '@p and @q are equal within $epsilon' );

This function tests for equality within a given value of epsilon. The test is
true if the absolute value of the difference between $p and $q is B<less than>
epsilon.  If the test is true, it prints an "OK" statement for use in testing.
If the test is not true, this function prints a failure report and diagnostic.
Epsilon must be non-zero.

The values to compare may be scalars or references to arrays.  If the values
are references to arrays, the comparison is done pairwise for each index value
of the array.  The pairwise comparison is recursive, so matrices may
be compared as well.

For example, this code sample compares two matrices:

    my @a = (   [ 3.14, 6.28 ],
                [ 1.41, 2.84 ]   );

    my @b = (   [ 3.14, 6.28 ],
                [ 1.42, 2.84 ]   );

    delta_within( \@a, \@b, 1e-6, 'compare @a and @b' );

The sample prints the following:

    not ok 1 - compare @a and @b
    # At [1][0]: 1.4100000 and 1.4200000 are not equal to within 0.000001

=cut

sub delta_within($$$;$) { ## no critic
    my ( $p, $q, $epsilon, $name ) = @_;
    croak "Value of epsilon to delta_within must be non-zero"
      if !defined($epsilon) || $epsilon == 0;
    {
        local $Test::Builder::Level = $Test::Builder::Level + 1;
        _delta_within( $p, $q, $epsilon, $name );
    }
}

sub _delta_within {
    my ( $p, $q, $epsilon, $name ) = @_;
    my ( $ok, $diag, @indices ) = _check( $p, $q, $epsilon, $name );
    if (@indices) {
        $diag = "At [" . join( "][", @indices ) . "]: $diag";
    }
    return $Test->ok( $ok, $name ) || $Test->diag($diag);
}

#--------------------------------------------------------------------------#
# delta_ok()
#--------------------------------------------------------------------------#

=head2 delta_ok

 delta_ok(  $p,  $q, '$p and $q are close enough to equal' );
 delta_ok( \@p, \@q, '@p and @q are close enough to equal' );

This function tests for equality within a default epsilon value.  See L</USAGE>
for details on changing the default.  Otherwise, this function works the same
as C<delta_within>.

=cut

sub delta_ok($$;$) { ## no critic
    my ( $p, $q, $name ) = @_;
    {
        local $Test::Builder::Level = $Test::Builder::Level + 1;
        _delta_within( $p, $q, undef, $name );
    }
}

#--------------------------------------------------------------------------#
# delta_not_ok()
#--------------------------------------------------------------------------#

=head2 delta_not_within

 delta_not_within(  $p,  $q, '$p and $q are different' );
 delta_not_within( \@p, \@q, $epsilon, '@p and @q are different' );

This test compares inequality in excess of a given value of epsilon. The test
is true if the absolute value of the difference between $p and $q is B<greater
than> epsilon.  For array or matrix comparisons, the test is true if I<any>
pair of values differs by more than epsilon.  Otherwise, this function works
the same as C<delta_within>.

=cut

sub delta_not_within($$$;$) { ## no critic
    my ( $p, $q, $epsilon, $name ) = @_;
    croak "Value of epsilon to delta_not_within must be non-zero"
      if !defined($epsilon) || $epsilon == 0;
    {
        local $Test::Builder::Level = $Test::Builder::Level + 1;
        _delta_not_within( $p, $q, $epsilon, $name );
    }
}

sub _delta_not_within($$$;$) { ## no critic
    my ( $p, $q, $epsilon, $name ) = @_;
    my ( $ok, undef, @indices ) = _check( $p, $q, $epsilon, $name );
    $ok = !$ok;
    my ( $ep, $dp ) = _ep_dp($epsilon);
    my $diag =
      defined($epsilon)
      ? sprintf( "Arguments are equal to within %.${ep}f", abs($epsilon) )
      : _diag_default();
    return $Test->ok( $ok, $name ) || $Test->diag($diag);
}

=head2 delta_not_ok

 delta_not_ok(  $p,  $q, '$p and $q are different' );
 delta_not_ok( \@p, \@q, '@p and @q are different' );

This function tests for inequality in excess of a default epsilon value.  See
L</USAGE> for details on changing the default.  Otherwise, this function works
the same as C<delta_not_within>.

=cut

sub delta_not_ok($$;$) { ## no critic
    my ( $p, $q, $name ) = @_;
    {
        local $Test::Builder::Level = $Test::Builder::Level + 1;
        _delta_not_within( $p, $q, undef, $name );
    }
}

=head1 SEE ALSO

=for :list
* L<Number::Tolerant>
* L<Test::Deep::NumberTolerant>

=cut

1;


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.025.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 6.17;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Compare the difference between numbers against a given tolerance",
  "AUTHOR" => "David Golden <dagolden\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.17"
  },
  "DISTNAME" => "Test-Number-Delta",
  "EXE_FILES" => [],
  "LICENSE" => "apache",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Test::Number::Delta",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Exporter" => 0,
    "Test::Builder" => 0,
    "strict" => 0,
    "vars" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::Builder::Tester" => "1.02",
    "Test::More" => 0
  },
  "VERSION" => "1.07",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => "6.17",
  "File::Spec" => 0,
  "Test::Builder" => 0,
  "Test::Builder::Tester" => "1.02",
  "Test::More" => 0,
  "strict" => 0,
  "vars" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



